### ISSUE TITLE
man: document explicitly that network-online.target is a one time thing, and doesn't follow online state

### ISSUE THREAD

**Issue by julian-klode** [systemd/systemd#5909]
### Submission type

  - [x] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

> 232

### Used distribution

Debian

### In case of bug report: Expected behaviour you didn't see

The network-online.target should be shut down when suspending and start again when resuming, so timers (or well, the services they activate, I guess) can reasonably specify Wants/After and have things work not only at boot, but also when the system was suspended.

### In case of bug report: Unexpected behaviour you saw

Target remains active all the time.

### In case of bug report: Steps to reproduce the problem

Suspend / resume.

**Comment by julian-klode** [comment]
I'm not entirely sure how this could work with the services, I guess the individual services wanted by the target also need appropriate settings?

**Comment by julian-klode** [comment]
Or, alternatively, maybe we could have some service internet-up.service that does not remain after exit and checks that a predefined internet location is reachable and times out after (let's say) 30 seconds or so? That would probably help too (it would also be a good idea even with a restarting network-online.target).

**Comment by poettering** [comment]
I am pretty sure we shouldn't try to make PID 1 a network state manager, we can only fail in such a goal. With "network-online.target" we do the absolutely minimum to make NFS work the way it traditionally worked, but not more. In addition we only provide a conceptual whitespace that people can fill with whatever they want.

If you want a more comprehensive management solution then I think you have to look for something else, for example networkd or nm, and their respective "wait-online" tools or so, which permit you to configure exactly what to wait for, which network-online.target really can't map...

We live in a highly dynamic networking world these days, with a variety of definitions what "online" means, and with services that can deal with different levels of online diffently well. Trying to expose this all unde the simplistic and binary "network-online.target" concept can never work.

Sorry if that's disappointing, but I really don#t think we should make n-o.t more than it is right now...
Sorry!

**Comment by hendriktews** [comment]
Lennart Poettering <notifications@github.com> writes:

> With "network-online.target" we do the absolutely minimum to
> make NFS work the way it traditionally worked, but not more.

After resume the network-online.target stays active, even if
network manager fails to connect again. How could NFS work in
such a situation?

The systemd documentation says

    network-online.target is a target that actively waits until the
    nework is "up", where the definition of "up" is defined by the
    network management software.

After resume systemd fails to fulfill this promise. The network
manager is in state disconnected with no connectivity, but
network-online.target stays active.

> I am pretty sure we shouldn't try to make PID 1 a network state
> manager, we can only fail in such a goal.

Sure, but maybe you could rely on the network management
software, as the documentation says? Otherwise I would suggest to
change the documentation to make clear that after a change in the
network configuration (especially after resume) discrepancies
between network-online.target and the state of the network
management software can occur.

Hendrik

**Comment by poettering** [comment]
> Otherwise I would suggest to change the documentation to make clear that after a change in the network configuration (especially after resume) discrepancies between network-online.target and the state of the network management software can occur.

True, we should make this clearer.

**Comment by poettering** [comment]
Documentation fix waiting in #7345.

**Comment by eMPee584** [comment]
@poettering: how is it that so many times, when (often somewhat inadvertent) users of your »grand vision™ solutions« chime in with their own (even legitimate) use cases, you're default reaction is like 

> nah, that's out of scope #final_poettering_verdict

Well at least four years after its introduction, network-online is finally documented to not actually mean the network is online, just that the networking stack has been fired up _once since boot_. Good I only wasted less than a single day finding this out...
For the record, for my use case, I resorted to planting a `/etc/NetworkManager/dispatcher.d/90autossh`:
```
#!/bin/sh -e
case "$2" in up|connectivity-change) if /usr/bin/nm-online -q; then systemctl reload autossh.service fi;; esac
```
But this very clearly needs a nicer systemd.unit™ solution. High hopes it will be technically possibly within less than a decade. 🌈🏭
